mshadow/dot_engine-inl.h

```diff
@@ -280,55 +281,75 @@ struct BLASEngine<cpu, float> {
-                          float beta, float *C, int ldc) {
+                          index_t m, index_t n, index_t k, float alpha,
+                          const float *A, index_t lda, const float *B, index_t ldb,
+                          float beta, float *C, index_t ldc) {
     cblas_sgemm(CblasColMajor, GetT(transa), GetT(transb),
                 m, n, k, alpha, A, lda, B, ldb, beta, C, ldc);
   }
   inline static void batched_gemm(Stream<cpu> *stream,
                                   bool transa, bool transb,
-                                  int m, int n, int k, float alpha,
-                                  const float *A, int lda, const float *B, int ldb,
-                                  float beta, float *C, int ldc, int batch_count,
+                                  index_t m, index_t n, index_t k, float alpha,
+                                  const float *A, index_t lda, const float *B, index_t ldb,
+                                  float beta, float *C, index_t ldc, index_t batch_count,
                                   float **workspace) {
 #if (MSHADOW_USE_MKL && INTEL_MKL_VERSION >= 20160000)
-  std::vector<int> p_m(batch_count, m);
-  std::vector<int> p_n(batch_count, n);
-  std::vector<int> p_k(batch_count, k);
-  std::vector<int> p_lda(batch_count, lda);
-  std::vector<int> p_ldb(batch_count, ldb);
-  std::vector<int> p_ldc(batch_count, ldc);
-  std::vector<float> p_alpha(batch_count, alpha);
-  std::vector<float> p_beta(batch_count, beta);
-  std::vector<const float*> pp_A;
-  std::vector<const float*> pp_B;
-  std::vector<float*> pp_C;
+    std::vector<int> p_m(batch_count, m);
+    std::vector<int> p_n(batch_count, n);
+    std::vector<int> p_k(batch_count, k);
+    std::vector<int> p_lda(batch_count, lda);
+    std::vector<int> p_ldb(batch_count, ldb);
+    std::vector<int> p_ldc(batch_count, ldc);
+    std::vector<float> p_alpha(batch_count, alpha);
+    std::vector<float> p_beta(batch_count, beta);
+    std::vector<const float*> pp_A;
+    std::vector<const float*> pp_B;
+    std::vector<float*> pp_C;
 
-  CBLAS_TRANSPOSE cblas_a_trans = GetT(transa);
-  CBLAS_TRANSPOSE cblas_b_trans = GetT(transb);
+    CBLAS_TRANSPOSE cblas_a_trans = GetT(transa);
+    CBLAS_TRANSPOSE cblas_b_trans = GetT(transb);
 
-  std::vector<int> p_group_sizeb(batch_count, batch_count);
-  std::vector<CBLAS_TRANSPOSE> p_transa(batch_count, cblas_a_trans);
-  std::vector<CBLAS_TRANSPOSE> p_transb(batch_count, cblas_b_trans);
+    std::vector<int> p_group_sizeb(batch_count, batch_count);
+    std::vector<CBLAS_TRANSPOSE> p_transa(batch_count, cblas_a_trans);
+    std::vector<CBLAS_TRANSPOSE> p_transb(batch_count, cblas_b_trans);
 
-  auto m_k = m * k;
-  auto k_n = k * n;
-  auto m_n = m * n;
 
-  for (int i = 0; i < batch_count; i++) {
-    pp_A.push_back(A + i * m_k);
-    pp_B.push_back(B + i * k_n);
-    pp_C.push_back(C + i * m_n);
-  }
+    int m_k = 0;
+    CHECK(mult_not_overflow(m,k, &m_k));
```

Comment:
Would a series call of this function cause runtime regression? Can we check this during debug mode?

Reply:
It's going to have some performance cost, this should be measured, but a set of divisions independent of tensor shapes shouldn't be too bad, it's basically O(1) perfomance penalty.